### PR TITLE
Let all functions be non-deterministic by default; explicitly set deterministic flag

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
~~The motivation of this PR are:~~
- ~~`Signature.Builder.features(Set)` overrode the given set with the set of features accumulated.~~
- ~~`deterministic` flag was set by default but it was not visible.~~
- ~~Most of the functions exist are `deterministic`, makes more sense to tag non-deterministic functions.~~

Update: As mentioned, https://github.com/crate/crate/pull/16174#issuecomment-2175362528, it is safer to mis-tag `deterministic` as `non-deterministic`. So setting all functions as `non-deterministic` by default. Also, updating `Signature.Builder.features(Set)` to union the existing set with the given set instead of overriding.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
This will make it easier to refactor https://github.com/crate/crate/pull/16133.

File: server/src/main/java/io/crate/metadata/functions/Signature.java [hunk 1]
```diff
@@ -416,7 +430,20 @@ public void writeTo(StreamOutput out) throws IOException {
             TypeSignature.toStream(typeSignature, out);
         }
         TypeSignature.toStream(returnType, out);
-        out.writeVInt(EnumSets.packToInt(features));
+        if (out.getVersion().onOrAfter(Version.V_5_8_0)) {
+            out.writeVInt(EnumSets.packToInt(features));
+        } else {
+            // before 5.8.0, the ordinal for Scalar.Feature.NON_DETERMINISTIC was Scalar.Feature.DETERMINISTIC
+            if (features.contains(Scalar.Feature.NON_DETERMINISTIC)) {
+                var copy = new HashSet<>(features);
```
Comment (matriv) [inline server/src/main/java/io/crate/metadata/functions/Signature.java:438]: double checking: it's because `features` is unmodifiable, right?

**Review by matriv — APPROVED** [review]
LGTM, thx!
(left minor comments)

File: server/src/main/java/io/crate/metadata/functions/Signature.java (hunk 1 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/metadata/functions/Signature.java:438]: Yes, that was the reason.

**Comment by jeeminso** [comment]
> > Most of the functions exist are deterministic, makes more sense to tag non-deterministic functions.
> 
> Afaik a function tagged as non-deterministic will just result in missed optimizations, while a function tagged as deterministic that isn't will result in incorrect behavior.
> 
> That makes the current existing default the safer choice. I'm not sure it's an improvement to flip that around?
> 
> (Would also be good to have up-front conversations about changes like this)

Hi @mfussenegger,  sorry I only saw this as a minor refactoring... I agree that `deterministic` should be the default. Could I revert the flipped enum and merge this? cc @matriv

File: server/src/main/java/io/crate/metadata/functions/Signature.java [hunk 2]
```diff
@@ -150,7 +150,7 @@ public static class Builder {
         private FunctionType kind;
         private List<TypeSignature> argumentTypes = Collections.emptyList();
         private TypeSignature returnType;
-        private Set<Scalar.Feature> features = Scalar.DETERMINISTIC_ONLY;
+        private Set<Scalar.Feature> features = Set.of();
```
Comment (jeeminso) [inline server/src/main/java/io/crate/metadata/functions/Signature.java:153]: NOTE: It used to be `deterministic` by default.

File: server/src/main/java/io/crate/expression/scalar/AgeFunction.java [hunk 3]
```diff
@@ -52,7 +52,7 @@ public static void register(Functions.Builder builder) {
                 NAME,
                 DataTypes.TIMESTAMP.getTypeSignature(),
                 DataTypes.INTERVAL.getTypeSignature()
-            ).withFeatures(EnumSet.of(Feature.NULLABLE)),
+            ).withFeatures(EnumSet.of(Feature.NON_DETERMINISTIC, Feature.NULLABLE)),
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/scalar/AgeFunction.java]: Thanks I can do that, but could I do that after this and https://github.com/crate/crate/pull/16133?

**Comment by jeeminso** [comment]
Could you please review/re-review this @matriv, @seut ? I made some changes regarding Jordi's review.(tested locally with c5c4ccfa5de6a14a13bae1bf46e5f6d66c756c84 to prevent regressions.)

**Comment by matriv** [comment]
I'm not sure adding `DETERMINISTIC` to all functions (even aggregations, hyperloglog, etc.) is correct.
As far as I understand this is used mainly in the normalizer, so if the flag is true (since now the feature is set everywhere) it will try to normalize functions that cannot be normalized anyway, I think that's a drawback.

Please wait for other's opinion as well.

**Review by matriv — APPROVED** [review]
Thx

**Comment by jeeminso** [comment]
Thanks for reviewing @matriv . Merging this since it's purely aesthetic changes.

**Comment by jeeminso** [comment]
retest this please

**Pull request merged**